training_data_loader.cpp

```diff
@@ -1079,31 +1079,72 @@ extern "C" {
 
 }
 
-/* benches */
-
-// #include <chrono>
-// #include <iostream>
-
-// int main(int argc, char** argv)
-// {
-//     if (argc < 2) {
-//         std::cerr << "Usage: " << argv[0] << " file1 [file2 ...]\n";
-//         return 1;
-//     }
-
-//     const char** files = const_cast<const char**>(&argv[1]);
-//     int file_count = argc - 1;
-
-//     auto stream = create_sparse_batch_stream("HalfKP", 1, file_count, files, 8192, true, false, 0, false, -1, 0);
-
-//     auto t0 = std::chrono::high_resolution_clock::now();
-//     for (int i = 0; i < 1000; ++i)
-//     {
-//         if (i % 100 == 0) std::cout << i << '\n';
-//         destroy_sparse_batch(stream->next());
-//     }
-//     auto t1 = std::chrono::high_resolution_clock::now();
-
-//     std::cout << (t1 - t0).count() / 1e9 << "s\n";
-//     return 0;
-// }
+#if defined(BENCH)
+
+/* benches
+   compile and run with:
+     g++ -g3 -O3 -DNDEBUG -DBENCH -march=native training_data_loader.cpp && ./a.out /path/to/binpack
+*/
+
+#include <chrono>
+#include <iostream>
+#include <iomanip>
+#include <fstream>
+#include <string>
+
+long long get_rchar_self() {
+    std::ifstream io_file("/proc/self/io");
+    std::string line;
+    while (std::getline(io_file, line)) {
+        if (line.rfind("rchar:", 0) == 0) {
+            return std::stoll(line.substr(6));
+        }
+    }
+    return -1; // Error or not found
+}
+
+int main(int argc, char** argv)
+{
+    if (argc < 2) {
+        std::cerr << "Usage: " << argv[0] << " file1 [file2 ...]\n";
+        return 1;
+    }
+
+    const char** files = const_cast<const char**>(&argv[1]);
+    int file_count = argc - 1;
+
+    const int concurrency = 64;
```

Comment:
maybe `std::min(64, std::thread::hardware_concurrency())` ?

Reply:
Even without the min... I agree